crates/prune/src/segments/account_history.rs

```diff
@@ -35,208 +37,341 @@ impl<DB: Database> Segment<DB> for AccountHistory {
-            done,
-            pruned: pruned_changesets + pruned_indices,
+            progress,
+            pruned,
             checkpoint: Some(PruneOutputCheckpoint {
-                block_number: Some(last_changeset_pruned_block),
+                block_number: last_pruned_block,
                 tx_number: None,
             }),
         })
     }
+
+    fn new_limiter_from_parent_scope_limiter(&self, limiter: &PruneLimiter) -> PruneLimiter {
+        PruneLimiterBuilder::floor_deleted_entries_limit_to_multiple_of(
+            limiter,
+            NonZeroUsize::new(2).unwrap(),
+        )
+        .build()
+    }
+}
+
+#[allow(missing_debug_implementations)]
+struct AccountHistoryTablesIter<'a, 'b, DB>
+where
+    DB: Database,
+{
+    provider: &'a DatabaseProviderRW<DB>,
+    limiter: &'b mut PruneLimiter,
+    last_pruned_block: Option<u64>,
+    to_block: u64,
+}
+
+impl<'a, 'b, DB> AccountHistoryTablesIter<'a, 'b, DB>
+where
+    DB: Database,
+{
+    fn new(
+        provider: &'a DatabaseProviderRW<DB>,
+        limiter: &'b mut PruneLimiter,
+        last_pruned_block: Option<u64>,
+        to_block: u64,
+    ) -> Self {
+        Self { provider, limiter, last_pruned_block, to_block }
+    }
+}
+
+impl<'a, 'b, DB> Iterator for AccountHistoryTablesIter<'a, 'b, DB>
+where
+    DB: Database,
+{
+    type Item = Result<Option<u64>, PrunerError>;
+    fn next(&mut self) -> Option<Self::Item> {
+        let Self { provider, limiter, last_pruned_block, to_block } = self;
+
+        if limiter.is_limit_reached() || Some(*to_block) == *last_pruned_block {
+            return None
+        }
+
+        let block_step =
+            if let Some(block) = *last_pruned_block { block + 1..=block + 2 } else { 0..=1 };
+
+        let next_up_last_pruned_block = Some(*block_step.start());
+        let mut last_pruned_block_changesets = None;
+        // todo: guarantee skip filter and delete callback are same for all header table types
+
+        let to_block = match provider.with_walker::<tables::AccountChangeSets, _, _>(
+            block_step.clone(),
+            |ref mut walker| {
+                provider.step_prune_range(walker, limiter, &mut |_| false, &mut |row| {
+                    last_pruned_block_changesets = Some(row.0)
+                })
+            },
+        ) {
+            Err(err) => return Some(Err(err.into())),
+            Ok(res) => if res.is_done() {
+                last_pruned_block_changesets
+            } else {
+                last_pruned_block_changesets.map(|block| block.saturating_sub(1))
+            }
+            .unwrap_or(*block_step.end()),
+        };
+
+        if let Err(err) = provider.with_cursor::<tables::AccountsHistory, _, _>(|ref mut cursor| {
+            step_prune_indices::<DB, _, _>(
+                cursor,
+                to_block,
+                limiter,
+                &|a, b| a.key == b.key,
+                &|key| ShardedKey::last(key.key),
+            )
```

Comment:
if I understand correctly, what happens here is that we prune one block in `AccountChangeSets`, and then delete this one block from all indices. If that's so, I think it'd be very slow because indices tables are large.

I'm not sure yet how to correctly solve this considering the timeout.